### ISSUE TITLE
[orderbook-v1] Is the distinction between bid and ask necessary?

### ISSUE THREAD

**Issue by maltekliemann** [zeitgeistpm/zeitgeist#1112]
It seems to me that the distinction between bid and ask in the implementation is unnecessary.

The only difference is in which asset is going in which direction. But in the end both are trades where one asset is traded against another. By removing this distinction from the `Order` object, we can avoid the match expressions which contain mostly duplicated code. It would also make benchmarking much more straightforward.

I'm not arguing for removing the enum `OrderSide`, by the way. I think the pallet's API is well-designed already. But there's no reason to use `OrderSide` in `Order`. Instead, `Order` should just have a `maker_asset` and `taker_asset`. If `maker_asset` is the base asset, it's a bid; otherwise, it's an ask.

**Comment by Chralt98** [comment]
> we can avoid the match expressions which contain mostly duplicated code.

I will update the code so that we have less duplications.

> Instead, Order should just have a maker_asset and taker_asset. If maker_asset is the base asset, it's a bid; otherwise, it's an ask.

I don't see how this simplifies our code. Can you give an example?

```rust
pub type OrderId = u128;

#[derive(Clone, Encode, Eq, Decode, MaxEncodedLen, PartialEq, RuntimeDebug, TypeInfo)]
pub enum OrderSide {
    Bid,
    Ask,
}

#[derive(Clone, Encode, Eq, Decode, MaxEncodedLen, PartialEq, RuntimeDebug, TypeInfo)]
pub struct Order<AccountId, Balance, MarketId: MaxEncodedLen> {
    pub market_id: MarketId,
    pub side: OrderSide,
    pub maker: AccountId,
    pub outcome_asset: Asset<MarketId>,
    pub base_asset: Asset<MarketId>,
    pub outcome_asset_amount: Balance,
    pub base_asset_amount: Balance,
}
```

to this?

```rust
pub type OrderId = u128;

#[derive(Clone, Encode, Eq, Decode, MaxEncodedLen, PartialEq, RuntimeDebug, TypeInfo)]
pub struct Order<AccountId, Balance, MarketId: MaxEncodedLen> {
    pub market_id: MarketId,
    pub maker: AccountId,
    pub taker_asset: Asset<MarketId>,
    pub maker_asset: Asset<MarketId>,
    pub outcome_asset_amount: Balance,
    pub base_asset_amount: Balance,
}
```

**Comment by maltekliemann** [comment]
The point is to eliminate the large `match` statement from `fill_order`. If you put the order side into the order, you probably won't be able to do that. I guess my point is that we probably don't even need the notion of bid/ask at all.

**Comment by maltekliemann** [comment]
Just to elaborate on this a little more. The whole distinction between bid and ask is based on the notion of a trading pair, such as KO/USD. Asks offer KO in exchange for USD, bids offer USD in exchange for KO. In the end, both do the same thing: Offer one asset in exchange for another. Hence, in the spirit of keeping things simple, there is no reason to even thing of them as different things unless the trading pair plays a _really_ important role (which it doesn't). One great way of looking at this is that each bid in the trading pair A1/A2 is an ask in the trading pair A2/A1.

In prediction markets, we of course think in terms of trading pairs OUTCOME/COLLATERAL, so bids are offers to buy outcomes and asks are offers to sell outcomes, but there's no reason to enforce this distinction in the code.

Long story short: Every order is fully defined by which asset it offers and which asset it expects in return. The way the code handles orders should be completely independent of whether or not the order is considered a bid or an ask.

Going back to your question which I neglected to answer, yeah, I think that's the way to go, except that `outcome_asset_amount` and `base_asset_amount` should be `taker_amount` and `maker_amount`. Let me know if you run into any trouble with this.
